File: app/src/core/model/infrastructure/building/academy/recruit-queue.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import { AcademyRecruitOrderImpl } from './recruit-order';
import type { AcademyRecruitQueue } from '@/types/core/infrastructure/academy';
import { InfrastructureQueueImpl } from '@/core/model/infrastructure/queue/queue';

export class AcademyRecruitQueueImpl extends InfrastructureQueueImpl<AcademyRecruitOrderImpl>
  implements AcademyRecruitQueue
{
  private constructor(queue: AcademyRecruitQueue) {
    const orders = queue.orders.map((it) => AcademyRecruitOrderImpl.create(it));
    super({ orders });
  }

  public static create(queue: AcademyRecruitQueue) {
    return new AcademyRecruitQueueImpl(queue);
  }
}

File: crates/nil-core/src/infrastructure/building/academy/recruit_queue.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use crate::continent::Coord;
use crate::decl_recruit_queue;
use crate::error::{Error, Result};
use crate::infrastructure::queue::{InfrastructureQueue, InfrastructureQueueOrder};
use crate::military::squad::{Squad, SquadSize};
use crate::military::unit::{AcademyUnitId, UnitBox};
use crate::resources::Resources;
use crate::resources::workforce::Workforce;
use serde::{Deserialize, Serialize};
use std::collections::VecDeque;
use std::num::NonZeroU32;
use strum::EnumIs;
use uuid::Uuid;

decl_recruit_queue!(Academy);
